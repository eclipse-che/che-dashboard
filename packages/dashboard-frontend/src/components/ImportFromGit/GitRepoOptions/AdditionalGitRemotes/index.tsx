/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import {
  ActionGroup,
  Button,
  ButtonVariant,
  Form,
  FormFieldGroup,
  FormSection,
} from '@patternfly/react-core';
import { PlusCircleIcon } from '@patternfly/react-icons';
import { isEqual } from 'lodash';
import React from 'react';

import AdditionalGitRemote from '@/components/ImportFromGit/GitRepoOptions/AdditionalGitRemotes/gitRemote';
import styles from '@/components/ImportFromGit/GitRepoOptions/AdditionalGitRemotes/index.module.css';
import { GitRemote } from '@/components/WorkspaceProgress/CreatingSteps/Apply/Devfile/getGitRemotes';

export type Props = {
  onChange: (remotes: GitRemote[] | undefined, isValid: boolean) => void;
  remotes: GitRemote[] | undefined;
};
export type State = {
  remotes: GitRemote[];
};

export class AdditionalGitRemotes extends React.PureComponent<Props, State> {
  private readonly callbacks: { getValidation?: () => boolean }[];

  constructor(props: Props) {
    super(props);

    const remotes = this.props.remotes ? [...this.props.remotes] : [];
    if (remotes.length === 0) {
      remotes.push({ name: '', url: '' });
    }
    this.state = {
      remotes,
    };
    this.callbacks = remotes.map(() => ({}));
  }

  public componentDidMount(): void {
    const remotes = this.props.remotes || [];
    if (remotes.length > 0) {
      this.props.onChange(this.state.remotes, this.getValidate());
    }
  }

  public componentDidUpdate(prevProps: Readonly<Props>, prevState: Readonly<State>): void {
    const remotes = this.props.remotes || [];
    const prevRemotes = prevProps.remotes || [];

    if (!isEqual(remotes, prevRemotes) && !isEqual(remotes, this.state.remotes)) {
      this.setState({
        remotes,
      });
    } else if (!isEqual(prevState.remotes, this.state.remotes)) {
      const validate = this.getValidate();
      this.props.onChange(this.state.remotes, validate);
    }
  }

  private getValidate(): boolean {
    const invalidIndex = this.callbacks.findIndex(
      callback => callback.getValidation && !callback.getValidation(),
    );
    return invalidIndex === -1;
  }

  private handleChange(index: number, remote: GitRemote): void {
    const remotes = [...this.state.remotes];
    remotes[index] = remote;

    this.setState({ remotes });
  }

  private handleAdd(): void {
    const remotes = [...this.state.remotes];
    remotes.push({ name: '', url: '' });

    this.setState({ remotes });
  }

  private handleRemove(index: number): void {
    const remotes = [...this.state.remotes];
    remotes.splice(index, 1);
    this.callbacks.splice(index, 1);

    this.setState({ remotes });
  }

  public buildGitRemotes(): React.JSX.Element {
    const { remotes } = this.state;
    this.callbacks.length = remotes.length;

    return (
      <>
        {remotes.map((remote: GitRemote, index: number) => {
          if (this.callbacks[index] === undefined) {
            this.callbacks[index] = {};
          }
          return (
            <AdditionalGitRemote
              key={'remote_' + index}
              remote={remote}
              onChange={(remote: GitRemote) => this.handleChange(index, remote)}
              callbacks={this.callbacks[index]}
              onDelete={() => this.handleRemove(index)}
            />
          );
        })}
      </>
    );
  }

  public render() {
    return (
      <Form isHorizontal={true}>
        <FormSection title="Additional Git Remotes" className={styles.gitRemotesTitle}>
          <FormFieldGroup className={styles.gitRemotesGroup}>
            {this.buildGitRemotes()}
            <ActionGroup style={{ marginTop: 0 }}>
              <Button
                onClick={() => this.handleAdd()}
                variant={ButtonVariant.link}
                style={{ textDecoration: 'none' }}
                isInline
                icon={<PlusCircleIcon />}
              >
                Add Remote
              </Button>
            </ActionGroup>
          </FormFieldGroup>
        </FormSection>
      </Form>
    );
  }
}
