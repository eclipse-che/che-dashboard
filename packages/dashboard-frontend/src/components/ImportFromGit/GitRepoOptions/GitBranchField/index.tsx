/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { FormGroup, TextInput } from '@patternfly/react-core';
import React from 'react';

export type Props = {
  onChange: (definition: string | undefined) => void;
  gitBranch: string | undefined;
};
export type State = {
  gitBranch: string | undefined;
};

export class GitBranchField extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      gitBranch: this.props.gitBranch,
    };
  }

  public shouldComponentUpdate(nextProps: Readonly<Props>, nextState: Readonly<State>): boolean {
    return (
      this.state.gitBranch !== nextState.gitBranch || this.props.gitBranch !== nextProps.gitBranch
    );
  }

  public componentDidUpdate(prevProps: Readonly<Props>): void {
    const { gitBranch } = this.props;
    if (prevProps.gitBranch !== gitBranch) {
      this.setState({ gitBranch });
    }
  }

  private handleChange(value: string) {
    let gitBranch: string | undefined = value.trim();
    gitBranch = gitBranch !== '' ? gitBranch : undefined;
    if (gitBranch !== this.state.gitBranch) {
      this.setState({ gitBranch: value });
      this.props.onChange(value);
    }
  }

  public render() {
    const gitBranch = this.state.gitBranch || '';

    return (
      <FormGroup label="Git Branch">
        <TextInput
          aria-label="Git Branch"
          placeholder="Enter the branch of the Git Repository"
          onChange={value => this.handleChange(value)}
          value={gitBranch}
        />
      </FormGroup>
    );
  }
}
