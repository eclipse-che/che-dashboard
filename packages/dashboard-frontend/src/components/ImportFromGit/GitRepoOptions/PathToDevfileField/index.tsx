/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { FormGroup, TextInput } from '@patternfly/react-core';
import React from 'react';

export type Props = {
  onChange: (devfilePath: string | undefined) => void;
  devfilePath: string | undefined;
};
export type State = {
  devfilePath: string | undefined;
};

export class PathToDevfileField extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      devfilePath: this.props.devfilePath,
    };
  }

  public shouldComponentUpdate(nextProps: Readonly<Props>, nextState: Readonly<State>): boolean {
    return (
      this.state.devfilePath !== nextState.devfilePath ||
      this.props.devfilePath !== nextProps.devfilePath
    );
  }

  public componentDidUpdate(prevProps: Readonly<Props>): void {
    const { devfilePath } = this.props;
    if (prevProps.devfilePath !== devfilePath) {
      this.setState({ devfilePath });
    }
  }

  private handleChange(value: string) {
    let devfilePath: string | undefined = value.trim();
    devfilePath = devfilePath !== '' ? devfilePath : undefined;
    if (devfilePath !== this.state.devfilePath) {
      this.setState({ devfilePath: value });
      this.props.onChange(value);
    }
  }

  public render() {
    const devfilePath = this.state.devfilePath || '';

    return (
      <FormGroup label="Path to Devfile">
        <TextInput
          aria-label="Path to Devfile"
          placeholder="Enter the relative path to the Devfile in the Git Repository"
          onChange={value => this.handleChange(value)}
          value={devfilePath}
        />
      </FormGroup>
    );
  }
}
