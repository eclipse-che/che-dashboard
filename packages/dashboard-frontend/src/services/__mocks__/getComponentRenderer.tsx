/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

export default function <T extends Array<unknown>>(
  getComponent: (..._args: T) => React.ReactElement,
): {
  createSnapshot: (...args: T) => renderer.ReactTestRenderer;
  renderComponent: (...args: T) => {
    reRenderComponent: (..._args: T) => void;
  };
} {
  return {
    createSnapshot: (...args: T) => {
      return renderer.create(getComponent(...args));
    },
    renderComponent: (...args: T) => {
      const res = render(getComponent(...args));
      return {
        reRenderComponent: (..._args: T) => {
          res.rerender(getComponent(..._args));
        },
      };
    },
  };
}
