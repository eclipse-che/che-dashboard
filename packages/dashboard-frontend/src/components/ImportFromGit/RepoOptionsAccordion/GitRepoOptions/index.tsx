/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Form } from '@patternfly/react-core';
import { isEqual } from 'lodash';
import React from 'react';

import { AdditionalGitRemotes } from '@/components/ImportFromGit/RepoOptionsAccordion/GitRepoOptions/AdditionalGitRemotes';
import { GitBranchField } from '@/components/ImportFromGit/RepoOptionsAccordion/GitRepoOptions/GitBranchField';
import { PathToDevfileField } from '@/components/ImportFromGit/RepoOptionsAccordion/GitRepoOptions/PathToDevfileField';
import { GitRemote } from '@/components/WorkspaceProgress/CreatingSteps/Apply/Devfile/getGitRemotes';

export type Props = {
  gitBranch: string | undefined;
  remotes: GitRemote[] | undefined;
  devfilePath: string | undefined;
  hasSupportedGitService: boolean;
  onChange: (
    gitBranch: string | undefined,
    remotes: GitRemote[] | undefined,
    devfilePath: string | undefined,
    isValid: boolean,
  ) => void;
};

export type State = {
  gitBranch: string | undefined;
  remotes: GitRemote[] | undefined;
  devfilePath: string | undefined;
  isValid: boolean;
};

export class GitRepoOptions extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      gitBranch: props.gitBranch,
      remotes: props.remotes,
      devfilePath: props.devfilePath,
      isValid: true,
    };
  }

  public componentDidUpdate(prevProps: Readonly<Props>): void {
    const { gitBranch, remotes, devfilePath } = this.props;

    if (gitBranch !== prevProps.gitBranch) {
      this.setState({ gitBranch });
    }

    if (!isEqual(remotes, prevProps.remotes)) {
      this.setState({ remotes });
    }

    if (devfilePath !== prevProps.devfilePath) {
      this.setState({ devfilePath });
    }
  }
  private handleGitBranch(gitBranch: string | undefined) {
    const { remotes, devfilePath, isValid } = this.state;

    this.setState({ gitBranch });
    this.props.onChange(gitBranch, remotes, devfilePath, isValid);
  }

  private handleRemotes(remotes: GitRemote[] | undefined, isValid: boolean) {
    const { gitBranch, devfilePath } = this.state;

    this.setState({ remotes, isValid });
    this.props.onChange(gitBranch, remotes, devfilePath, isValid);
  }

  private handleDevfilePath(devfilePath: string | undefined) {
    const { gitBranch, remotes, isValid } = this.state;

    this.setState({ devfilePath });
    this.props.onChange(gitBranch, remotes, devfilePath, isValid);
  }
  public render() {
    const { hasSupportedGitService } = this.props;
    const { gitBranch, remotes, devfilePath } = this.state;
    return (
      <Form isHorizontal={true} onSubmit={e => e.preventDefault()}>
        {hasSupportedGitService && (
          <GitBranchField
            onChange={gitBranch => this.handleGitBranch(gitBranch)}
            gitBranch={gitBranch}
          />
        )}
        <AdditionalGitRemotes
          onChange={(remotes: GitRemote[] | undefined, isValid: boolean) =>
            this.handleRemotes(remotes, isValid)
          }
          remotes={remotes}
        />
        <PathToDevfileField
          onChange={devfilePath => this.handleDevfilePath(devfilePath)}
          devfilePath={devfilePath}
        />
      </Form>
    );
  }
}
