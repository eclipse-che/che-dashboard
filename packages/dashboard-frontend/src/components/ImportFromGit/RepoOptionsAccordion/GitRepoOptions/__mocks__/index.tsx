/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/components/ImportFromGit/RepoOptionsAccordion/GitRepoOptions';

export class GitRepoOptions extends React.PureComponent<Props> {
  public render() {
    const { gitBranch, remotes, devfilePath, hasSupportedGitService, onChange } = this.props;

    return (
      <div>
        <div>Git Repo Options</div>
        <div data-testid="git-repo-options">{`${gitBranch}, ${JSON.stringify(remotes)}, ${devfilePath}, ${hasSupportedGitService}`}</div>
        <button
          onClick={() =>
            onChange(
              'newBranch',
              [{ name: 'test-updated', url: 'http://test' }],
              'newDevfilePath',
              true,
            )
          }
        >
          Git Repo Options Change
        </button>
      </div>
    );
  }
}
