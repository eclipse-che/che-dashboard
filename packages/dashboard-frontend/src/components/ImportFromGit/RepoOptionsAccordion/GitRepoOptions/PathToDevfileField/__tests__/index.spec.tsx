/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import userEvent from '@testing-library/user-event';
import React from 'react';

import { PathToDevfileField } from '@/components/ImportFromGit/RepoOptionsAccordion/GitRepoOptions/PathToDevfileField';
import getComponentRenderer, { screen } from '@/services/__mocks__/getComponentRenderer';

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

const mockOnChange = jest.fn();

describe('PathToDevfileField', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('snapshot', () => {
    const snapshot = createSnapshot();
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test('devfile path preset value', () => {
    renderComponent('preset-devfile-path');

    const input = screen.getByRole('textbox');

    expect(input).toHaveValue('preset-devfile-path');
  });

  test('devfile path change', () => {
    renderComponent();

    const input = screen.getByRole('textbox');

    const devfilePath = 'new-devfile-path';
    userEvent.paste(input, devfilePath);

    expect(mockOnChange).toHaveBeenNthCalledWith(1, devfilePath);

    userEvent.clear(input);
    expect(mockOnChange).toHaveBeenNthCalledWith(2, '');
  });
});

function getComponent(devfilePath?: string) {
  return <PathToDevfileField devfilePath={devfilePath} onChange={mockOnChange} />;
}
