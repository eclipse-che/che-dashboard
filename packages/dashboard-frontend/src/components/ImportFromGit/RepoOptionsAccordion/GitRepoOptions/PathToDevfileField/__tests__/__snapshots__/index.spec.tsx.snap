// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PathToDevfileField snapshot 1`] = `
<div
  className="pf-c-form__group"
>
  <div
    className="pf-c-form__group-label"
  >
    <label
      className="pf-c-form__label"
    >
      <span
        className="pf-c-form__label-text"
      >
        Path to Devfile
      </span>
    </label>
     
  </div>
  <div
    className="pf-c-form__group-control"
  >
    <input
      aria-invalid={false}
      aria-label="Path to Devfile"
      className="pf-c-form-control"
      data-ouia-component-id="OUIA-Generated-TextInputBase-1"
      data-ouia-component-type="PF4/TextInput"
      data-ouia-safe={true}
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Enter the relative path to the Devfile in the Git Repository"
      required={false}
      type="text"
      value=""
    />
  </div>
</div>
`;
